ios/PurchasesHybridCommon/PurchasesHybridCommonSwift/CommonFunctionality.swift

```diff
@@ -0,0 +1,513 @@
+                completion(eligibilityByProductId.mapValues { [
+                    "status": $0.status,
+                    "description": $0.description
+                ]
+                })
+            }
+        }
+
+
+    @objc static func getProductInfo(_ productIds: [String], completionBlock: @escaping([[String: Any]]) -> Void) {
+        Purchases.shared.products(productIds) { products in
+            let productDictionaries = products.map { $0.rc_dictionary }
+            completionBlock(productDictionaries)
+        }
+    }
+
+    @objc(paymentDiscountForProductIdentifier:discount:completionBlock:)
+    static func paymentDiscount(for productIdentifier: String,
+                                discountIdentifier: String?,
+                                completion: @escaping ([String: Any]?, ErrorContainer?) -> Void) {
+        guard #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) else {
+            completion(nil, nil)
+            return
+        }
+
+        product(with: productIdentifier) { product in
+            guard let product = product else {
+                completion(nil, productNotFoundError(description: "Couldn't find product", userCancelled: false))
+                return
+            }
+
+            guard let discountToUse = discount(with: discountIdentifier, for: product) else {
+                completion(nil, productNotFoundError(description: "Couldn't find discount", userCancelled: false))
+                return
+            }
+
+            let paymentDiscountCompletion: (SKPaymentDiscount?, Error?) -> Void = { discount, error in
+                guard let discount = discount else {
+                    if let error = error {
+                        completion(nil, ErrorContainer(error: error, extraPayload: [:]))
+                    } else {
+                        let error = NSError(domain: Purchases.ErrorDomain,
+                                            code: Purchases.ErrorCode.unknownError.rawValue,
+                                            userInfo: [NSLocalizedDescriptionKey: description])
+
+                        completion(nil, ErrorContainer(error: error, extraPayload: [:]))
+                    }
+                    return
+                }
+                discountsByProductIdentifier[discount.timestamp.stringValue] = discount
+                completion(discount.rc_dictionary, nil)
+            }
+
+            Purchases.shared.paymentDiscount(for: discountToUse,
+                                             product: product,
+                                             completion: paymentDiscountCompletion)
+
+        }
+    }
+
+}
+
+// MARK: Subscriber attributes
+@objc public extension CommonFunctionality {
+
+    @objc static func setAttributes(_ attributes: [String: Any]) {
+        Purchases.shared.setAttributes(attributes.mapValues { $0 as? String ?? "" })
+    }
+
+    @objc static func setEmail(_ email: String?) {
```

Comment:
the method signature doesn't allow for NSNull. I think this is fine. The old code did check if the value was NSNull, but the compiler should actually issue a warning and disallow it in the first place, so I think this extra check wasn't needed.